src/btree/bt_page.c

```diff
@@ -418,20 +454,178 @@ __wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint32
     return (ret);
 }
 
+/*
+ * __wt_col_fix_read_auxheader --
+ *     Read the auxiliary header following the bitmap data, if any. This code is used by verify and
+ *     needs to be accordingly careful. It is also used by mainline reads so it must also not crash
+ *     or print.
+ */
+int
+__wt_col_fix_read_auxheader(
+  WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, WT_COL_FIX_AUXILIARY_HEADER *auxhdr)
+{
+    WT_BTREE *btree;
+    uint32_t bitmapsize, auxheaderoffset;
+    const uint8_t *raw, *end;
+
+    btree = S2BT(session);
+
+    /*
+     * Figure where the auxiliary header is. It is always immediately after the bitmap data,
+     * regardless of whether the page is full.
+     */
+    bitmapsize = __bitstr_size(dsk->u.entries * btree->bitcnt);
+    auxheaderoffset = WT_PAGE_HEADER_BYTE_SIZE(btree) + bitmapsize;
+
+    /*
+     * If the auxiliary header is past the in-memory page size, there's no auxiliary data. If
+     * there's at least one byte past the bitmap data, check whether it's zero. If that's zero,
+     * there's no auxiliary data. (We are guaranteed that any allocation slop that we might be
+     * looking at is all zeros.) Set everything to zero and return.
+     */
+    if (auxheaderoffset >= dsk->mem_size || *(raw = (uint8_t *)dsk + auxheaderoffset) == 0) {
+        auxhdr->version = WT_COL_FIX_VERSION_NIL;
+        auxhdr->entries = 0;
+        auxhdr->offset = 0;
+        return (0);
+    }
+
+    /* Remember the end of the page for easy computation of maximum lengths. */
+    end = (uint8_t *)dsk + dsk->mem_size;
+
+    /*
+     * The on-disk header is a 1-byte version, a uint32 with the number of entries, and a uint32
+     * that says how many waste bytes there are between the header and the entries.
+     */
+    auxhdr->version = raw[0];
+    raw++;
+    {
+        uint32_t entries, waste;
+
+        if (raw + sizeof(entries) > end)
+            return (EINVAL);
+        memcpy(&entries, raw, sizeof(entries));
+        raw += sizeof(entries);
+        if (raw + sizeof(waste) > end)
+            return (EINVAL);
+        memcpy(&waste, raw, sizeof(waste));
+        raw += sizeof(waste);
+
+#ifdef WORDS_BIGENDIAN
+        entries = __wt_bswap32(entries);
+        waste = __wt_bswap32(waste);
+#endif
```

Comment:
Yeah. I could change them back to packed, but the net result would be making every page two bytes longer (since in general you need to reserve 5 for a packed integer, and... wait, that's not true. Can make every page three bytes _shorter_.

(Because we need to start writing time windows before we know how much data there is, they have to go where there's enough room for a header after a full page of data. Currently that's 9 bytes, and naively with packed integers it becomes 11, _but_ in this case the waste space entry is small and will fit in one byte, which makes the header at most 6. So as long zero packs into 1 byte, which it does, and values less than 11 - 6 = 5 pack into no more than 2, which they do, reserving 6 bytes is sufficient -- if the bitmap data is 5 bytes short of full or smaller there's 11 bytes for a maximal header, if it's less than that a 6 or even 7 byte header will fit, and if the page is full the header will be 6 bytes.)

Lemme try it to make sure the code doesn't become horrifying. Not entirely trivial so won't be in the first batch of changes, though.

Reply:
Done. It's not horrifying.